### PR TITLE
Pass cancellation tokens and tasks to OperationCanceledException and TaskCanceledException

### LINKED ISSUE #2100: "FSharpDocumentDiagnosticAnalyzer throws TaskCanceledException" warning
image

### PR DESCRIPTION
**Pull request by saul**
This is exhibiting itself in #2100. Roslyn has a try/catch in [ComputeProjectDiagnosticAnalyzerDiagnosticsAsync](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs,71ad559976c39f3c) to see whether the cancellation came from the same cancellation token as the one that was passed into the analyzer.

The problem is Async isn't passing in the Task or CancelationToken to the exceptions, which C# does.

Fixes #2100

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
Wow. It's deep.

**Comment by saul** [comment]
@vasily-kirichenko the only ones that matter for #2100 are the TaskCanceledException. The OperationCanceledException changes are just for completeness ;)

**Comment by dsyme** [comment]
Wow, I never knew.  Great fix!

File: src/fsharp/FSharp.Core/control.fs [hunk 1]
```diff
@@ -1541,8 +1541,8 @@ namespace Microsoft.FSharp.Control
             let continuation (completedTask : Task<_>) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
-                        if useCcontForTaskCancellation then args.aux.ccont(new OperationCanceledException())
-                        else args.aux.econt (ExceptionDispatchInfo.Capture(new TaskCanceledException()))
+                        if useCcontForTaskCancellation then args.aux.ccont(new OperationCanceledException(args.aux.token))
```
Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core/control.fs]: @saul @dsyme this is incorrect. You are passing the async cancellation token which may not be related to the cancellation of the task we're binding on.

File: src/fsharp/FSharp.Core/control.fs [hunk 2]
```diff
@@ -1555,8 +1555,8 @@ namespace Microsoft.FSharp.Control
             let continuation (completedTask : Task) : unit =
                 args.aux.trampolineHolder.Protect((fun () ->
                     if completedTask.IsCanceled then
-                        if useCcontForTaskCancellation then args.aux.ccont (new OperationCanceledException())
-                        else args.aux.econt (ExceptionDispatchInfo.Capture(new TaskCanceledException()))
+                        if useCcontForTaskCancellation then args.aux.ccont (new OperationCanceledException(args.aux.token))
```
Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core/control.fs]: Same here.

**Comment by saul** [comment]
@eiriktsarpalis if you have a fix please open a PR

File: src/fsharp/FSharp.Core/control.fs (hunk 1 above)
Comment (eiriktsarpalis) [inline src/fsharp/FSharp.Core/control.fs]: @saul @matthid this line should be reverted to the original (unless there is a generic way to obtain a cancellation token from arbitrary tasks which I'm not aware of)

Comment (matthid) [inline src/fsharp/FSharp.Core/control.fs]: Ok added in https://github.com/Microsoft/visualfsharp/pull/3255

Comment (dsyme) [inline src/fsharp/FSharp.Core/control.fs]: @eiriktsarpalis @matthid What is this were  ``if args.aux.token.IsCancellationRequested then new OperationCanceledException(args.aux.token) else new OperationCanceledException() ``?   

I suspect a lot of corresponding C# code would do the same thing: if a cancellation continuation needs to be converted into a OCE, then you would attach the current cancellation token in scope to the OCE, at least on the condition that that the current cancellation token in scope is in the cancelled state,

**Pull request merged**